infrastructure/gcp/keep-test/google-functions/keep-faucet/issue-grant.js

```diff
@@ -39,72 +38,117 @@ const web3 = new Web3(keepContractOwnerProvider, null, web3Options)
-        Token grant failed, your account has the maximum testnet KEEP allowed.
-        You can manage your token grants at: https://dashboard.test.keep.network
-        If you have questions find us on Discord: https://discord.gg/jqxBU4m\n`)
     } else {
-      const grantData = Buffer.concat([
-        Buffer.from(granteeAccount.substr(2), 'hex'),
-        web3.utils.toBN(unlockingDuration).toBuffer('be', 32),
-        web3.utils.toBN(start).toBuffer('be', 32),
-        web3.utils.toBN(cliff).toBuffer('be', 32),
-        Buffer.from(revocable ? '01' : '00', 'hex'),
-      ])
+      const granteeAccount = account
+      const unlockingDuration = web3.utils.toBN(0)
+      const start = web3.utils.toBN(Math.floor(Date.now() / 1000))
+      const cliff = web3.utils.toBN(0)
+      const revocable = false
+      const tokens = web3.utils.toBN(300000)
+      console.log(`Fetching existing balance for account [${granteeAccount}]...`)
+      const grantBalanceString = await tokenGrant.methods
+        .balanceOf(granteeAccount)
+        .call()
+      const grantBalance = web3.utils.toBN(grantBalanceString)
+      console.log(`Existing balance for account [${granteeAccount}] is [${grantBalance}].`)
+      const grantAmount = tokens.mul(tokenDecimalMultiplier)
 
-      await keepToken.methods
-        .approveAndCall(tokenGrant.address, grantAmount, grantData)
-        .send({ from: keepContractOwnerAddress })
+      if (grantBalance.gte(grantAmount)) {
+        console.warn(
+          `[${granteeAccount}] requested grant while at limit. Balance: [${grantBalance}].`,
+        )
+        return response.status(400).send(`
+          Token grant failed: your account has the maximum testnet KEEP allowed.\n
+          You can manage your token grants at: https://dashboard.test.keep.network\n
+          If you have questions, you can find us on Discord: https://discord.gg/jqxBU4m\n`
+        )
+      } else {
+        console.log(
+          `Submitting grant for [${grantAmount}] to [${granteeAccount}]...`,
+        )
+        const grantData = Buffer.concat([
+          Buffer.from(granteeAccount.substr(2), 'hex', 20),
+          unlockingDuration.toBuffer('be', 32),
+          start.toBuffer('be', 32),
+          cliff.toBuffer('be', 32),
+          Buffer.from(revocable ? '01' : '00', 'hex'),
+          Buffer.from(permissiveStakingPolicyAddress.substr(2), 'hex', 20)
+        ])
 
-      console.log(
-        `Created grant for ${web3.utils.toBN(
-          grantAmount,
-        )} to: ${granteeAccount}`,
-      )
-      response.send(`
-        Created token grant with ${web3.utils.toBN(
-          grantAmount,
-        )} KEEP for account: ${granteeAccount}
-        You can manage your token grants at: https://dashboard.test.keep.network
-        You can find us on Discord at: https://discord.gg/jqxBU4m\n`)
+        console.log("Test submission...")
+        // Try calling; if this throws, we'll have a proper error message thanks
```

Comment:
This is neat.

Reply:
I both agree and desperately hate it lol.